packages/typespec-client-generator-core/src/decorators.ts

```diff
@@ -1029,26 +1035,71 @@ const clientInitializationKey = createStateSymbol("clientInitialization");
 export const $clientInitialization: ClientInitializationDecorator = (
   context: DecoratorContext,
   target: Namespace | Interface,
-  options: Model,
+  options: Type,
   scope?: LanguageScopes,
 ) => {
-  setScopedDecoratorData(
-    context,
-    $clientInitialization,
-    clientInitializationKey,
-    target,
-    options,
-    scope,
-  );
+  if (options.kind === "Model") {
+    if (options.properties.get("initializedBy")) {
+      const value = options.properties.get("initializedBy")!.type;
+
+      const isValidValue = (value: number): boolean => value === 1 || value === 2;
+
+      if (value.kind === "EnumMember") {
+        if (typeof value.value !== "number" || !isValidValue(value.value)) {
+          reportDiagnostic(context.program, {
+            code: "invalid-initialized-by",
+            format: { message: "Please use `InitializedBy` enum to set the value." },
+            target: target,
+          });
+          return;
+        }
+      } else if (value.kind === "Union") {
+        for (const variant of value.variants.values()) {
+          if (
+            variant.type.kind !== "EnumMember" ||
+            typeof variant.type.value !== "number" ||
+            !isValidValue(variant.type.value)
+          ) {
+            reportDiagnostic(context.program, {
+              code: "invalid-initialized-by",
+              format: { message: "Please use `InitializedBy` enum to set the value." },
+              target: target,
+            });
+            return;
+          }
+        }
+      }
+    }
+
+    setScopedDecoratorData(
+      context,
+      $clientInitialization,
+      clientInitializationKey,
+      target,
+      options,
+      scope,
+    );
+  }
 };
 
+/**
+ * Get `SdkInitializationType` for namespace or interface. The info is from `@clientInitialization` decorator.
+ *
+ * @param context
+ * @param entity namespace or interface which represents a client
+ * @returns
+ * @deprecated This function is deprecated. Use `getClientInitializationOptions` instead.
+ */
 export function getClientInitialization(
   context: TCGCContext,
   entity: Namespace | Interface,
 ): SdkInitializationType | undefined {
-  const model = getScopedDecoratorData(context, clientInitializationKey, entity);
-  if (!model) return model;
-  const sdkModel = getSdkModel(context, model);
+  let options = getScopedDecoratorData(context, clientInitializationKey, entity);
+  if (options === undefined) return undefined;
+  // backward compatibility
+  if (options.properties.get("initializedBy")) return undefined;
```

Comment:
why `return undefined` when properties have `initializedBy`?

I assume it should still get to the value of `parameters`, if available (kind of ignore `initializedBy`)?

Or, it should be `undefined` only if `initializedBy=parent`?

Reply:
fixed.